cpp/ql/src/Diagnostics/FailedExtractions.qll

```diff
@@ -0,0 +1,121 @@
+import cpp
+
+/**
+ * The class of errors upon we mark a file as non-successfully extracted.
+ */
+class ReportableError extends Diagnostic {
+  ReportableError() {
+    (
+      this instanceof CompilerDiscretionaryError or
+      this instanceof CompilerError or
+      this instanceof CompilerCatastrophe
+    ) and
+    // If the extractor encounters an error in a compilation, it always emits a
+    // catch-all diagnostic "There was an error during this compilation", to ensure
+    // that the error makes it to the database.
+    // This error doesn't have a file path attached to it, and is thus
+    // useless for us to report. Furthermore, in the common case, we will have a
+    // proper diagnostic for this error we can show.
+    // Instead, we synthesize `TUnknownError` if this is the only error that we can show to the user.
+    not this.getFile().getAbsolutePath() = ""
+  }
+}
+
+private newtype TExtractionError =
+  TReportableError(ReportableError err) or
+  TCompilationFailed(Compilation c, File f) {
+    f = c.getAFileCompiled() and not c.normalTermination()
+  } or
+  // Report generic extractor errors only if we haven't seen any other error-level diagnostic
+  TUnknownError(CompilerError err) { not exists(ReportableError e) }
+
+/**
+ * Superclass for the extraction error hierarchy.
+ */
+class ExtractionError extends TExtractionError {
+  string toString() { none() }
+
+  /** Gets the error message for this error. */
+  string getErrorMessage() { none() }
+
+  /** Gets the file this error occured in. */
+  File getFile() { none() }
+
+  /** Gets the location this error occured in. */
+  Location getLocation() { none() }
+
+  /** Gets the SARIF severity of this error. */
+  int getSeverity() {
+    // Unfortunately, we can't distinguish between errors and fatal errors in SARIF,
+    // so all errors have severity 2.
+    result = 2
+  }
+}
+
+/**
+ * An irrecoverable extraction failure, where extraction was unable to finish.
```

Comment:
I would have said "unrecoverable", but I'm not sure either is necessarily wrong. Let's make a decision so that we're consistent going forward.

Reply:
A search for `unrecoverable` vs `irrecoverable` in our code base agrees with you.
I have no feelings either way, so I'm changing it to unrecoverable.

Reply:
I think it'd say _non-recoverable_, but I'm sure either is fine.

Reply:
Or we could say _fatal_ and _non-fatal_. Recoverability mostly makes sense when we're talking about a parser, I think.

Reply:
Sorry, I don't want to dive into a terminology discussion. What we have seems fine.

Reply:
Searching for all 3 options in our source code yields that we use all options, but `unrecoverable` is the most popular option, so I'm going with that.
If there's a push at some point to standardize, we can change it later, it's not exactly part of our public API.